pyanaconda/payload/manager.py

```diff
@@ -31,93 +32,88 @@
 
-    This class defines several states, and events can be triggered upon
-    reaching a state. Depending on whether a state has already been reached
-    when a listener is added, the event code may be run in either the
-    calling thread or the payload thread. The event code will block the
-    payload thread regardless, so try not to run anything that takes a long
-    time.
+    This class defines several states (PayloadState enum), and events can
+    be triggered upon reaching a state. Depending on whether a state has
+    already been reached when a listener is added, the event code may be
+    run in either the calling thread or the payload thread. The event code
+    will block the payload thread regardless, so try not to run anything
+    that takes a long time.
 
-    All states except STATE_ERROR are expected to happen linearly, and adding
+    All states except ERROR are expected to happen linearly, and adding
     a listener for a state that has already been reached or passed will
     immediately trigger that listener. For example, if the payload thread is
-    currently in STATE_GROUP_MD, adding a listener for STATE_NETWORK will
-    immediately run the code being added for STATE_NETWORK.
+    currently in DOWNLOADING_GROUP_METADATA, adding a listener for
+    WAITING_NETWORK will immediately run the code being added
+    for WAITING_NETWORK.
 
     The payload thread data should be accessed using the payloadMgr object,
     and the running thread can be accessed using threadMgr with the
     THREAD_PAYLOAD constant, if you need to wait for it or something. The
-    thread should be started using payloadMgr.restartThread.
+    thread should be started using payloadMgr.restart_thread.
     """
-
-    STATE_START = 0
-    # Waiting on storage
-    STATE_STORAGE = 1
-    # Waiting on network
-    STATE_NETWORK = 2
-    # Verify repository availability
-    STATE_TEST_AVAILABILITY = 3
-    # Downloading package metadata
-    STATE_PACKAGE_MD = 4
-    # Downloading group metadata
-    STATE_GROUP_MD = 5
-    # All done
-    STATE_FINISHED = 6
-
-    # Error
-    STATE_ERROR = -1
-
     # Error strings
     ERROR_SETUP = N_("Failed to set up installation source")
     ERROR_MD = N_("Error downloading package metadata")
 
     def __init__(self):
         self._event_lock = threading.Lock()
         self._event_listeners = {}
-        self._thread_state = self.STATE_START
+        self._thread_state = PayloadState.STARTED
         self._error = None
 
         # Initialize a list for each event state
-        for event_id in range(self.STATE_ERROR, self.STATE_FINISHED + 1):
-            self._event_listeners[event_id] = []
+        for name, value in PayloadState.__members__.items():
+            self._event_listeners[PayloadState(value)] = []
 
     @property
     def error(self):
         return _(self._error)
 
-    def addListener(self, event_id, func):
+    def add_listener(self, event_id, func):
```

Comment:
Wow, I didn't know about this code. It looks like a weird implementation of signals :-)

Reply:
Yeah, if I'm not mistaken this is much older than signals.